heka/sandbox/filters/firefox_aggregator.lua

```diff
@@ -0,0 +1,193 @@
+-- This Source Code Form is subject to the terms of the Mozilla Public
+-- License, v. 2.0. If a copy of the MPL was not distributed with this
+-- file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+--[[
+Collects the cbufd, txt, and json output from multiple instances of an upstream
+Firefox sandbox filter (the filters should all be the same version at least with
+respect to their cbuf output). The purpose is to recreate the view at a larger
+scope in each level of the aggregation
+i.e., host view -> datacenter view -> service level view.
+
+Config:
+
+- enable_delta (bool, optional, default false)
+    Specifies whether or not this aggregator should generate cbuf deltas.
+
+- anomaly_config(string) - (see :ref:`sandbox_anomaly_module`)
+    A list of anomaly detection specifications.  If not specified no anomaly
+    detection/alerting will be performed.
+
+- preservation_version (uint, optional, default 0)
+    If `preserve_data = true` is set in the SandboxFilter configuration, then
+    this value should be incremented every time the `enable_delta`
+    configuration is changed to prevent the plugin from failing to start
+    during data restoration.
+
+*Example Heka Configuration*
+
+.. code-block:: ini
+
+    [PluginNameAggregator]
+    type = "SandboxFilter"
+    message_matcher = "Logger =~ /^PluginName_/ && (Fields[payload_type] == 'cbufd' || Fields[payload_type] == 'json' || Fields[payload_type] == 'txt')"
+    ticker_interval = 60
+    filename = "lua_filters/firefox_aggregator.lua"
+    preserve_data = true
+--]]
+_PRESERVATION_VERSION = read_config("preservation_version") or 0
+
+require "cjson"
+local alert     = require "alert"
+local agg       = require "agg"
+local annotation= require "annotation"
+local anomaly   = require "anomaly"
+local cbufd     = require "cbufd"
+require "circular_buffer"
+
+local enable_delta = read_config("enable_delta") or false
+local anomaly_config = anomaly.parse_config(read_config("anomaly_config"))
+local last_update = 0
+local MAX_TTL = 24*60*60*1e9
+
+cbufs = {}
+hosts = {}
+
+local function init_cbuf(payload_name, h)
+    local cb = circular_buffer.new(h.rows, h.columns, h.seconds_per_row, enable_delta)
+    for i,v in ipairs(h.column_info) do
+        cb:set_header(i, v.name, v.unit, v.aggregation)
+    end
+    annotation.set_prune(payload_name, h.rows * h.seconds_per_row * 1e9)
+
+    cbufs[payload_name] = {cb = cb, last_update = 0}
+    return cbufs[payload_name]
+end
+
+local function update_cbuf(cb, data)
+    for i,v in ipairs(data) do
+        for col, value in ipairs(v) do
+            if value == value then -- NaN test, only aggregrate numbers
+                local n, u, agg = cb:get_header(col)
```

Comment:
If anything we should probably rename (agg) it is much too generic for a global level module or remove it all together since it is only used in one place.